x-pack/plugins/apm/server/lib/service_map/get_service_anomalies.ts

```diff
@@ -0,0 +1,158 @@
+import { APM_ML_JOB_GROUP_NAME } from '../../../common/ml_job_constants';
+
+type ApmMlJobCategory = NonNullable<ReturnType<typeof getApmMlJobCategory>>;
+const getApmMlJobCategory = (
+  mlJob: AnomalyDetectionJob,
+  serviceNames: string[]
+) => {
+  const apmJobGroups = mlJob.groups.filter(
+    (groupName) => groupName !== APM_ML_JOB_GROUP_NAME
+  );
+  if (apmJobGroups.length === mlJob.groups.length) {
+    // ML job missing "apm" group name
+    return;
+  }
+  const [serviceName] = intersection(apmJobGroups, serviceNames);
+  if (!serviceName) {
+    // APM ML job service was not found
+    return;
+  }
+  const [transactionType] = apmJobGroups.filter(
+    (groupName) => groupName !== serviceName
+  );
+  if (!transactionType) {
+    // APM ML job transaction type was not found.
+    return;
+  }
+  return { jobId: mlJob.job_id, serviceName, transactionType };
+};
+
+export type ServiceAnomalies = PromiseReturnType<typeof getServiceAnomalies>;
+
+export async function getServiceAnomalies(
+  options: IEnvOptions,
+  serviceNames: string[]
+) {
+  const { start, end, ml } = options.setup;
+
+  if (!ml || serviceNames.length === 0) {
+    return [];
+  }
+
+  const { jobs: apmMlJobs } = await ml.anomalyDetectors.jobs('apm');
+  const apmMlJobCategories = apmMlJobs
+    .map((job) => getApmMlJobCategory(job, serviceNames))
+    .filter(
+      (apmJobCategory) => apmJobCategory !== undefined
+    ) as ApmMlJobCategory[];
+  const apmJobIds = apmMlJobs.map((job) => job.job_id);
+  const rangeQuery = { range: rangeFilter(start, end, 'timestamp') };
+  const params = {
+    body: {
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            { term: { result_type: 'bucket' } },
+            {
+              terms: {
+                job_id: apmJobIds,
+              },
+            },
+            rangeQuery,
+          ],
+        },
+      },
+      aggs: {
+        jobs: {
+          terms: { field: 'job_id', size: apmJobIds.length },
+          aggs: {
+            top_score_hits: {
+              top_hits: {
+                sort: [{ anomaly_score: { order: 'desc' as const } }],
+                _source: ['anomaly_score', 'timestamp'],
+                size: 1,
+              },
+            },
+          },
+        },
+      },
+    },
+  };
+
+  const response = await ml.mlSystem.mlAnomalySearch(params);
+  const anomalyScores: Array<{
+    jobId: string;
+    anomalyScore: number;
+    timestamp: number;
+  }> = response.aggregations.jobs.buckets.map((jobBucket: any) => {
+    const jobId = jobBucket.key;
+    const bucketSource = jobBucket.top_score_hits.hits.hits?.[0]?._source;
+    return {
+      jobId,
+      anomalyScore: bucketSource.anomaly_score,
+      timestamp: bucketSource.timestamp,
+    };
+  });
+  const anomalyModelValuePromises = anomalyScores.map(
+    ({ jobId, timestamp }) => {
+      return (async () => {
+        try {
+          const modelPlotResponse = await ml.mlSystem.mlAnomalySearch({
```

Comment:
It looks like this is invoked inside a nested iteration. How many times is it called? Preferably we should be able to get anomalies with a single es request. is that no longer the case?

Reply:
it fetches the matching `model_plot` bucket by timestamp. and it fetches each one in a separate request. There wasn't a way to do it all in one request unfortunately, but this search is super fast since there's no scoring, and only returns the 1 hit.

Reply:
One alternative is to do the fetch when the user clicks on on the service to open the popover, and return it along with the other metrics.

Reply:
After going down this path, i realized that we no longer needed to use the `bucket` & `model_plot` result types in the anomalies query, So i returned to using the previous `record` type, which includes the `actual` & `typical` values in the same document. This prevents the need to make any extra requests, so this issue is now resolved.